lnwallet/rpcwallet/remote_signer_client.go

```diff
@@ -0,0 +1,977 @@
+	requestTimeout time.Duration) (*OutboundClient, error) {
+
+	if walletServer == nil || signerServer == nil {
+		return nil, errors.New("sub-servers cannot be nil when using " +
+			"an outbound remote signer")
+	}
+
+	if streamFeeder == nil {
+		return nil, errors.New("streamFeeder cannot be nil")
+	}
+
+	return &OutboundClient{
+		walletServer:    walletServer,
+		signerServer:    signerServer,
+		streamFeeder:    streamFeeder,
+		requestTimeout:  requestTimeout,
+		quit:            make(chan struct{}),
+		retryTimeout:    defaultRetryTimeout,
+		maxRetryTimeout: defaultMaxRetryTimeout,
+	}, nil
+}
+
+// Start starts the remote signer client. The function will continuously try to
+// setup a connection to the configured watch-only node, and retry to connect if
+// the connection fails until we Stop the remote signer client.
+func (r *OutboundClient) Start() error {
+	r.wg.Add(1)
+
+	// We'll continuously try setup a connection to the watch-only node, and
+	// retry to connect if the connection fails until we Stop the remote
+	// signer client.
+	go func() {
+		defer r.wg.Done()
+
+		for {
+			err := r.run()
+			if err != nil {
+				log.Errorf("Remote signer client error: %v",
+					err)
+			}
+
+			select {
+			case <-r.quit:
+				return
+			default:
+				log.Infof("Will retry to connect to "+
+					"watch-only node in: %v",
+					r.retryTimeout)
+
+				// Backoff before retrying to connect to the
+				// watch-only node.
+				select {
+				case <-r.quit:
+					return
+				case <-time.After(r.retryTimeout):
+				}
+			}
+
+			log.Infof("Retrying to connect to watch-only node")
+
+			// Increase the retry timeout by 50% for every retry.
+			r.retryTimeout = time.Duration(float64(r.retryTimeout) *
+				retryMultiplier)
+
+			// But cap the retryTimeout at r.maxRetryTimeout
+			if r.retryTimeout > r.maxRetryTimeout {
+				r.retryTimeout = r.maxRetryTimeout
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops the remote signer client.
+func (r *OutboundClient) Stop() error {
+	if r.stopped.Swap(true) {
+		return errors.New("remote signer client is already shut down")
+	}
+
+	log.Info("Remote signer client shutting down")
+
+	// Ensure that no new Run goroutines can start when we've initiated
+	// the stopping of the remote signer client.
+	r.wgMu.Lock()
+
+	if r.streamFeeder != nil {
+		r.streamFeeder.Stop()
+	}
+
+	close(r.quit)
+
+	r.wgMu.Unlock()
+
+	r.wg.Wait()
+
+	log.Debugf("Remote signer client shut down")
+
+	return nil
+}
+
+// run creates a new stream to the watch-only node, and starts processing and
+// responding to the sign requests that are sent over the stream. The function
+// will continuously run until the remote signer client is either stopped or
+// the stream errors.
+func (r *OutboundClient) run() error {
+	r.wgMu.Lock()
+
+	select {
+	case <-r.quit:
+		r.wgMu.Unlock()
+		return ErrShuttingDown
+	default:
+	}
+
+	r.wgMu.Unlock()
+
+	streamCtx, cancel := context.WithCancel(context.Background())
+
```

Comment:
run should take a context & not derive a fresh one

Reply:
I changed the `run` function so that it takes in passed `context`, and then place the creation of the fresh context within the `Start` function instead. Ideally the Start function should take a passed context that's used as well, but as no other server in `server.go` does so currently, I didn't include that change (but I expect that you'll change that in your upcoming context PR?). If you want me to add that change already here, so that the new context is created out in `server.go` so that it becomes easier for you to change that in your PR, please let me know :).